### PR TITLE
Pivotal ID # 177267343: Validate EU-ToxRisk Submissions

### PR DESCRIPTION
**Pull request by jhoanmanuelms**
https://www.pivotaltracker.com/story/show/177267343

- Implement collection validators
- Update the submission submitter to use them

Delivers # 177267343

### CODE REVIEW

File: commons/commons-model-extended-mapping/src/main/kotlin/ebi/ac/uk/extended/mapping/to/ToSubmission.kt [hunk 1]
```diff
@@ -18,11 +22,14 @@ fun ExtSubmission.toSimpleSubmission(): Submission = Submission(
 )
 
 private fun ExtSubmission.getSubmissionAttributes(): List<Attribute> {
-    val subAttrs = attributes.map { it.toAttribute() }.toMutableSet()
+    val subAttrs =
```
Comment (Juan-EBI) [inline commons/commons-model-extended-mapping/src/main/kotlin/ebi/ac/uk/extended/mapping/to/ToSubmission.kt:25]: why are we filtering it here?

Comment (jhoanmanuelms) [reply commons/commons-model-extended-mapping/src/main/kotlin/ebi/ac/uk/extended/mapping/to/ToSubmission.kt:25]: https://github.com/EBIBioStudies/biostudies-backend-services/pull/352/files/7b1cd4da93596bcedb0b382c16ffa1886d527ecf#diff-9f69c64127e65171eccdde20f28b77b1545af6262592db23b1512fade38f528cL16

File: commons/commons-model-extended-mapping/src/main/kotlin/ebi/ac/uk/extended/mapping/to/ToSubmission.kt [hunk 2]
```diff
@@ -18,11 +21,14 @@ fun ExtSubmission.toSimpleSubmission(): Submission = Submission(
...
+        attributes
```
Comment (Juan-EBI) [inline commons/commons-model-extended-mapping/src/main/kotlin/ebi/ac/uk/extended/mapping/to/ToSubmission.kt:25]: I think we should create a private method `getpageTabAttributes`  to make clear why we remove the validator.

File: commons/commons-test/src/main/kotlin/ebi/ac/uk/test/ExtSubmissionFactory.kt [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package ebi.ac.uk.test
+
+import ebi.ac.uk.extended.model.ExtProcessingStatus.PROCESSED
+import ebi.ac.uk.extended.model.ExtSection
+import ebi.ac.uk.extended.model.ExtSubmission
+import ebi.ac.uk.extended.model.ExtSubmissionMethod.PAGE_TAB
+import java.time.OffsetDateTime
+import java.time.ZoneOffset.UTC
+
+val basicExtSubmission = ExtSubmission(
```
Comment (Juan-EBI) [inline commons/commons-test/src/main/kotlin/ebi/ac/uk/test/ExtSubmissionFactory.kt:10]: we have create many of this, eventually we will need to put in common and use the same one

File: submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/service/ParentInfoService.kt [hunk 4]
```diff
@@ -2,26 +2,49 @@ package ac.uk.ebi.biostd.submission.service
...
+    fun executeCollectionValidators(submission: ExtSubmission) = submission.isCollection.ifFalse {
+        submission
+            .collections
+            .mapNotNull { collectionInfo(it.accNo).validator }
+            .forEach { validate(it, submission) }
+    }
+
+    private fun collectionInfo(accNo: String) = queryService.getBasicCollection(accNo)
+
+    private fun validate(validator: String, submission: ExtSubmission) =
```
Comment (Juan-EBI) [inline submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/service/ParentInfoService.kt:44]: this method is not doing what is said that is does making logic harder to follow, it does not validate but it gets the validation bean and validate (it is also doing to two operations and not a single one)

File: submission/submitter/src/test/kotlin/ac/uk/ebi/biostd/submission/validator/collection/EuToxRiskValidatorTest.kt [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        val error = assertThrows<CollectionValidationException> { testInstance.validate(submission) }
+        assertThat(error.message).isEqualTo(
+            "The submission doesn't comply with the collection requirements. Errors: [$EXCEL_FILE_REQUIRED]")
+    }
+
+    private fun setUpRestTemplate(response: EuToxRiskValidatorResponse): CapturingSlot<HttpEntity<FileSystemResource>> {
+        val requestSlot = slot<HttpEntity<FileSystemResource>>()
+
+        every {
+            restTemplate.postForObject(testUrl, capture(requestSlot), EuToxRiskValidatorResponse::class.java)
```
Comment (Juan-EBI) [inline submission/submitter/src/test/kotlin/ac/uk/ebi/biostd/submission/validator/collection/EuToxRiskValidatorTest.kt:94]: I think we can use `.postForObject<EuToxRiskValidatorResponse` can not we? as in the normal code.

File: submission/submitter/src/test/kotlin/ac/uk/ebi/biostd/submission/validator/collection/EuToxRiskValidatorTest.kt [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            files = listOf(left(ExtFile("test.txt", textFile)))
+        ))
+
+        val error = assertThrows<CollectionValidationException> { testInstance.validate(submission) }
+        assertThat(error.message).isEqualTo(
+            "The submission doesn't comply with the collection requirements. Errors: [$EXCEL_FILE_REQUIRED]")
+    }
+
+    private fun setUpRestTemplate(response: EuToxRiskValidatorResponse): CapturingSlot<HttpEntity<FileSystemResource>> {
+        val requestSlot = slot<HttpEntity<FileSystemResource>>()
```
Comment (Juan-EBI) [inline submission/submitter/src/test/kotlin/ac/uk/ebi/biostd/submission/validator/collection/EuToxRiskValidatorTest.kt:91]: IMO we should declare this as a field and in each test:
```
  every {  restTemplate.postForObject<>(testUrl, capture(requestSlot)) } } returns response
```
I think way when one is reviewing the test does not need to check what `setUpRestTemplate` does as the setup is part of each test. Also I think we do not need type  I think compiler with infer it.

**Review by Juan-EBI — APPROVED** [review]
LGTM please check comments before merge

File: commons/commons-test/src/main/kotlin/ebi/ac/uk/test/ExtSubmissionFactory.kt (hunk 3 above)
Comment (jhoanmanuelms) [reply commons/commons-test/src/main/kotlin/ebi/ac/uk/test/ExtSubmissionFactory.kt:10]: That's actually what I did with this change. I placed this one in commons-test and used it from everywhere else

**Pull request merged**